### PR TITLE
PeerMetrics: interpolate results when inserting a new peer

### LINKED ISSUE #3861: Initial score for new peers entering the churn leader board

### PR DESCRIPTION
**Pull request by coot**
# Description

This PR makes sure that newly added peers to `PeerMetrics` are never in 20% of worst performing peers.  Currently the churn interval and number of records kept by `PeerMetrics` are in sync, peers added just before the churn event can be removed as they might have too few results.  In the future the churn interval and the number of records kept by `PeerMetrics` might diverge, and thus it's more important to make sure that newly added peers are not immediately removed.

This PR solves this problem by keeping track of average results when a peer joined the leader board and is using linear regression to adjust peer results for the leader board period where the peer was absent from the leader board.

Fixes #3861

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [x] New tests are added if needed and existing tests are updated
    - [ ] If this branch changes Consensus and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../ouroboros-consensus/docs/interface-CHANGELOG.md)
    - [ ] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [ ] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

**Review by bolt12 — APPROVED** [review]
Very nice!

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 1]
```diff
@@ -231,6 +231,20 @@ data ArgumentsExtra m = ArgumentsExtra {
       -- is using @TIME_WAIT@.
       --
     , daTimeWaitTimeout :: DiffTime
+
+      -- | Churn interval between churn events in deadline mode.  A small fuzz
+      -- is added (max 10 minutes) so that not all nodes churn at the same time.
+      --
+      -- By default it is set to 3300 seconds.
+      --
+    , daDaedlineChurnInterval   :: DiffTime
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:240]:
```suggestion
    , daDeadlineChurnInterval   :: DiffTime
```

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 2]
```diff
@@ -606,6 +620,8 @@ runM Interfaces
        , daReadUseLedgerAfter
        , daProtocolIdleTimeout
        , daTimeWaitTimeout
+       , daDaedlineChurnInterval
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:623]:
```suggestion
       , daDeadlineChurnInterval
```

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 3]
```diff
@@ -879,6 +895,8 @@ runM Interfaces
                           Async.withAsync
                           (Governor.peerChurnGovernor
                             dtTracePeerSelectionTracer
+                            daDaedlineChurnInterval
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:898]:
```suggestion
                            daDeadlineChurnInterval
```

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 4]
```diff
@@ -1026,6 +1044,8 @@ runM Interfaces
                                   Async.withAsync
                                     (Governor.peerChurnGovernor
                                       dtTracePeerSelectionTracer
+                                      daDaedlineChurnInterval
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:1047]:
```suggestion
                                      daDeadlineChurnInterval
```

File: ouroboros-network/test/Test/Ouroboros/Network/Diffusion/Node.hs [hunk 5]
```diff
@@ -265,12 +265,14 @@ run blockGeneratorArgs limits ni na tracersExtra =
...
-      , Diff.P2P.daReadUseLedgerAfter   = aReadUseLedgerAfter na
-      , Diff.P2P.daProtocolIdleTimeout  = aProtocolIdleTimeout na
-      , Diff.P2P.daTimeWaitTimeout      = aTimeWaitTimeout na
+      { Diff.P2P.daPeerSelectionTargets  = aPeerSelectionTargets na
+      , Diff.P2P.daReadLocalRootPeers    = aReadLocalRootPeers na
+      , Diff.P2P.daReadPublicRootPeers   = aReadPublicRootPeers na
+      , Diff.P2P.daReadUseLedgerAfter    = aReadUseLedgerAfter na
+      , Diff.P2P.daProtocolIdleTimeout   = aProtocolIdleTimeout na
+      , Diff.P2P.daTimeWaitTimeout       = aTimeWaitTimeout na
+      , Diff.P2P.daDaedlineChurnInterval = 3300
```
Comment (bolt12) [inline ouroboros-network/test/Test/Ouroboros/Network/Diffusion/Node.hs:275]:
```suggestion
      , Diff.P2P.daDeadlineChurnInterval = 3300
```

**Review by bolt12 — APPROVED** [review]
Looks good! Just minor typo

**Comment by coot** [comment]
bors merge

**Pull request merged**